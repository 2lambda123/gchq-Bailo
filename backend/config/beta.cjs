module.exports = {
  experimental: {
    v2: true,
  },
}
