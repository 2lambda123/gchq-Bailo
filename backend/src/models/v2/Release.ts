import { Document, model, Schema } from 'mongoose'
import MongooseDelete from 'mongoose-delete'

// This interface stores information about the properties on the base object.
// It should be used for plain object representations, e.g. for sending to the
// client.
export interface ReleaseInterface {
  modelId: string
  modelCardVersion: number

  name: string
  semver: string
  notes: string

  minor: boolean
  draft: boolean

  files: Array<string>
  images: Array<string>

  deleted: boolean

  createdAt: Date
  updatedAt: Date
}

// The doc type includes all values in the plain interface, as well as all the
// properties and functions that Mongoose provides.  If a function takes in an
// object from Mongoose it should use this interface
export type ReleaseDoc = ReleaseInterface & Document<any, any, ReleaseInterface>

const ReleaseSchema = new Schema<ReleaseInterface>(
  {
    modelId: { type: String, required: true },
    modelCardVersion: { type: Number, required: true },

    name: { type: String, required: true },
    semver: { type: String, required: true },
    notes: { type: String, required: true },

    minor: { type: Boolean, required: true },
    draft: { type: Boolean, required: true },

    files: [{ type: String }],
    images: [{ type: String }],
  },
  {
    timestamps: true,
    collection: 'v2_releases',
  }
)

ReleaseSchema.plugin(MongooseDelete, { overrideMethods: 'all', deletedBy: true, deletedByType: Schema.Types.ObjectId })

const ReleaseModel = model<ReleaseInterface>('v2_Release', ReleaseSchema)

export default ReleaseModel
