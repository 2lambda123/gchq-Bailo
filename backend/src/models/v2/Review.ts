import { Document, model, Schema } from 'mongoose'
import MongooseDelete from 'mongoose-delete'

import { ReviewKind, ReviewKindKeys } from '../../types/v2/enums.js'

export const Decision = {
  RequestChanges: 'request_changes',
  Approve: 'approve',
} as const
export type DecisionKeys = (typeof Decision)[keyof typeof Decision]

interface ReviewResponse {
  user: string
  decision: DecisionKeys
  comment: string
}

// This interface stores information about the properties on the base object.
// It should be used for plain object representations, e.g. for sending to the
// client.
export interface ReviewInterface {
  semver?: string
  modelId?: string
  kind: ReviewKindKeys

  role: string

  responses: Array<ReviewResponse>

  createdAt: Date
  updatedAt: Date
}

// The doc type includes all values in the plain interface, as well as all the
// properties and functions that Mongoose provides.  If a function takes in an
// object from Mongoose it should use this interface
export type ReviewDoc = ReviewInterface & Document<any, any, ReviewInterface>

const ReviewSchema = new Schema<ReviewInterface>(
  {
    semver: {
      type: String,
      required: function (this: ReviewInterface): boolean {
        return this.kind === ReviewKind.Release
      },
    },
    modelId: { type: String, required: true },
    kind: { type: String, enum: Object.values(ReviewKind), required: true },

    role: { type: String, required: true },

    responses: [
      {
        user: { type: String, required: true },
        decision: { type: String, enum: Object.values(Decision), required: true },
        comment: { type: String, required: false },
      },
    ],
  },
  {
    timestamps: true,
    collection: 'v2_reviews',
  }
)

ReviewSchema.plugin(MongooseDelete, {
  overrideMethods: 'all',
  deletedBy: true,
  deletedByType: String,
})

const ReviewModel = model<ReviewInterface>('v2_Review', ReviewSchema)

export default ReviewModel
