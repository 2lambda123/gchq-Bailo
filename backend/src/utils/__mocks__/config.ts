const config = {
  app: {
    protocol: '',
    host: '',
    port: 3000,
  },
  s3: {
    buckets: {
      uploads: 'uploads',
      registry: 'registry',
    },
  },
  connectors: {
    authentication: {
      kind: 'silly',
    },
    audit: {
      kind: 'silly',
    },
    authorisation: {
      kind: 'basic',
    },
  },
  log: {
    level: 'debug',
  },
  registry: {
    connection: {
      internal: 'https://localhost:5000',
    },
  },
  oauth: {
    enabled: false,
  },
  experimental: {
    v2: true,
  },
}

export default config
