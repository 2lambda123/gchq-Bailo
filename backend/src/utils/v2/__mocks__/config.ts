const config = {
  connectors: {
    authentication: {
      kind: 'silly',
    },
    authorisation: {
      kind: 'silly',
    },
  },
  log: {
    level: 'debug',
  },
  registry: {
    connection: {
      internal: 'https://localhost:5000',
    },
  },
}

export default config
