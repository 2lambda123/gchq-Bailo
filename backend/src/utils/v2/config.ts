import bunyan from 'bunyan'
import _config from 'config'

export interface Config {
  connectors: {
    user: {
      kind: 'silly' | string
    }

    authorisation: {
      kind: 'silly' | string
    }
  }

  smtp: {
    enabled: boolean

    connection: {
      host: string
      port: number
      secure: boolean
      auth: {
        user: string
        pass: string
      }
      tls: {
        rejectUnauthorized: boolean
      }
    }

    from: string
  }

  log: {
    level: bunyan.LogLevel
  }
}

const config: Config = _config.util.toObject()
export default config
