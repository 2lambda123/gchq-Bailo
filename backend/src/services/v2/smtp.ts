import nodemailer, { Transporter } from 'nodemailer'

import config from '../../utils/v2/config.js'
import { GenericError } from '../../utils/v2/error.js'
import log from './log.js'

let transporter: undefined | Transporter = undefined

export async function sendEmail(to: string, subject: string, body: string, html?: string) {
  console.log(config)
  if (!config.smtp.enabled) {
    log.info({ subject, to }, 'Not sending email due to SMTP disabled')
    return
  }

  if (!transporter) {
    transporter = nodemailer.createTransport({
      host: config.smtp.connection.host,
      port: config.smtp.connection.port,
      secure: config.smtp.connection.secure,
      auth: config.smtp.connection.auth,
      tls: config.smtp.connection.tls,
    })
  }

  try {
    const info = await transporter.sendMail({
      from: config.smtp.from, // sender address
      to,
      subject,
      text: body,
      html,
    })
    log.info({ messageId: info.messageId }, 'Email sent')
  } catch (err) {
    throw GenericError(500, 'Error Sending email notification', { to, body, internal: err })
  }
}
