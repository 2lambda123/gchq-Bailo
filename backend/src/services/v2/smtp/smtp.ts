import nodemailer, { Transporter } from 'nodemailer'

import authorisation from '../../../connectors/v2/authorisation/index.js'
import { ReleaseDoc } from '../../../models/v2/Release.js'
import { ReviewRequestDoc } from '../../../models/v2/ReviewRequest.js'
import config from '../../../utils/v2/config.js'
import log from '../log.js'
import { ReleaseReviewRequestEmail } from './templates/releaseReviewRequest.js'
import { IEmailTemplate } from './templates/baseEmailTemplate.js'
import { emailDeploymentOwnersOnVersionDeletion } from '../../deployment.js'
import { getReleaseName } from '../release.js'

const appBaseUrl = `${config.app.protocol}://${config.app.host}:${config.app.port}`
let transporter: undefined | Transporter = undefined

export async function requestReviewForRelease(entity: string, reviewRequest: ReviewRequestDoc, release: ReleaseDoc) {
  if (!config.smtp.enabled) {
    log.info('Not sending email due to SMTP disabled')
    return
  }

  const releaseName = getReleaseName(release)
  const userInfoList = await Promise.all(await authorisation.getUserInformationList(entity))
  const sendEmailResponses = userInfoList.map(async (userInfo) => {
    const email = new ReleaseReviewRequestEmail()
    email.setTo(userInfo.email)
    email.setSubject(releaseName, reviewRequest.role)
    email.setText(releaseName, reviewRequest.kind, release.modelId, appBaseUrl, release.createdBy)
    email.setHtml(releaseName, reviewRequest.kind, release.modelId, appBaseUrl, release.createdBy)
    return await sendEmail(email)
  })
  await Promise.all(sendEmailResponses)
}

async function sendEmail(email: IEmailTemplate) {
  if (!transporter) {
    transporter = nodemailer.createTransport({
      host: config.smtp.connection.host,
      port: config.smtp.connection.port,
      secure: config.smtp.connection.secure,
      auth: config.smtp.connection.auth,
      tls: config.smtp.connection.tls,
    })
  }

  try {
    const info = await transporter.sendMail({
      from: config.smtp.from, // sender address
      to: email.to,
      subject: email.subject,
      text: email.text,
      html: email.html,
    })
    log.info({ messageId: info.messageId }, 'Email sent')
  } catch (err) {
    const content = { to: email.to, subject: email.subject, text: email.text }
    log.warn(`Unable to send email`, content)
    return Promise.reject(`Unable to send email: ${JSON.stringify(content)}`)
  }
}
