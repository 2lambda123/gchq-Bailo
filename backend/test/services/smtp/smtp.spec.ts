import { describe, expect, test, vi } from 'vitest'

import Release from '../../../src/models/v2/Release.js'
import Review from '../../../src/models/v2/Review.js'
import config from '../../../src/utils/v2/config.js'
import { requestReviewForRelease } from '../../../src/services/v2/smtp/smtp.js'

vi.mock('../../../src/utils/v2/config.js', () => {
  return {
    __esModule: true,
    default: {
      app: {
        protocol: '',
        host: '',
        port: 3000,
      },

      smtp: {
        enabled: true,

        connection: {
          host: 'localhost',
          port: 1025,
          secure: false,
          auth: undefined,
          tls: {
            rejectUnauthorized: false,
          },
        },

        from: '"Bailo 📝" <bailo@example.org>',
      },
    },
  }
})

const logMock = vi.hoisted(() => ({
  info: vi.fn(),
  warn: vi.fn(),
}))
vi.mock('../../../src/services/v2/log.js', async () => ({
  default: logMock,
}))

const transporterMock = vi.hoisted(() => {
  return {
    sendMail: vi.fn(() => ({ messageId: 123 })),
  }
})
const nodemailerMock = vi.hoisted(() => ({
  createTransport: vi.fn(() => transporterMock),
}))
vi.mock('nodemailer', async () => ({
  default: nodemailerMock,
}))

const authorisationMock = vi.hoisted(() => ({
  getUserInformationList: vi.fn(() => [Promise.resolve({ email: 'email@email.com' })]),
}))
vi.mock('../../../src/connectors/v2/authorisation/index.js', async () => ({ default: authorisationMock }))

const releaseEmailMock = vi.hoisted(() => {
  const obj: any = {
    from: '"Bailo 📝" <bailo@example.org>',
    to: 'email@email.com',
    subject: 'subject',
    text: 'text',
    html: 'html',
  }

  obj.setSubject = vi.fn()
  obj.setTo = vi.fn()
  obj.setText = vi.fn()
  obj.setHtml = vi.fn()

  const model: any = vi.fn(() => obj)
  Object.assign(model, obj)

  return model
})
vi.mock('../../../src/services/v2/smtp/templates/releaseReview.js', () => ({
  ReleaseReviewEmail: releaseEmailMock,
}))

describe('services > smtp > smtp', () => {
  test('that an email is sent', async () => {
    await requestReviewForRelease('user:user', new Review(), new Release())

    expect(transporterMock.sendMail.mock.calls.at(0)).toMatchSnapshot()
  })

  test('that an email is not sent when disabled in config', async () => {
    vi.spyOn(config, 'smtp', 'get').mockReturnValueOnce({
      enabled: false,
      connection: {
        host: 'localhost',
        port: 1025,
        secure: false,
        auth: { user: '', pass: '' },
        tls: {
          rejectUnauthorized: false,
        },
      },
      from: '"Bailo 📝" <bailo@example.org>',
    })

    await requestReviewForRelease('user:user', new Review(), new Release())

    expect(transporterMock.sendMail).not.toBeCalled()
  })

  test('that sendEmail is called for each member of a group entity', async () => {
    authorisationMock.getUserInformationList.mockReturnValueOnce([
      Promise.resolve({ email: 'member1@email.com' }),
      Promise.resolve({ email: 'member2@email.com' }),
    ])

    await requestReviewForRelease('group:group1', new Review(), new Release())

    expect(transporterMock.sendMail.mock.calls).toMatchSnapshot()
  })

  test('that we log when an email cannot be sent', async () => {
    authorisationMock.getUserInformationList.mockReturnValueOnce([
      Promise.resolve({ email: 'member1@email.com' }),
      Promise.resolve({ email: 'member2@email.com' }),
    ])
    transporterMock.sendMail.mockRejectedValueOnce('Failed to send email')

    const result: Promise<void> = requestReviewForRelease('user:user', new Review(), new Release())
    expect(result).rejects.toThrowError(`Unable to send email`)
  })
})
