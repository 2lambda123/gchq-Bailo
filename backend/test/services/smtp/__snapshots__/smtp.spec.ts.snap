// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > smtp > smtp > that an email is sent after a response for a an access request review 1`] = `undefined`;

exports[`services > smtp > smtp > that an email is sent after a response for a release review 1`] = `undefined`;

exports[`services > smtp > smtp > that an email is sent for Access Request Reviews 1`] = `
[
  {
    "from": "\\"Bailo 📝\\" <bailo@example.org>",
    "html": "html",
    "subject": "subject",
    "text": "text",
    "to": "email@email.com",
  },
]
`;

exports[`services > smtp > smtp > that an email is sent for Release Reviews 1`] = `
[
  {
    "from": "\\"Bailo 📝\\" <bailo@example.org>",
    "html": "html",
    "subject": "subject",
    "text": "text",
    "to": "email@email.com",
  },
]
`;

exports[`services > smtp > smtp > that sendEmail is called for each member of a group entity 1`] = `
[
  [
    {
      "from": "\\"Bailo 📝\\" <bailo@example.org>",
      "html": "html",
      "subject": "subject",
      "text": "text",
      "to": "member1@email.com",
    },
  ],
  [
    {
      "from": "\\"Bailo 📝\\" <bailo@example.org>",
      "html": "html",
      "subject": "subject",
      "text": "text",
      "to": "member2@email.com",
    },
  ],
]
`;
